data/model/oci/manifest.py

```diff
@@ -392,6 +321,131 @@ def _create_manifest(
     return CreatedManifest(manifest=manifest, newly_created=True, labels_to_apply=labels_to_apply)
 
 
+def _build_blob_map(
+    repository_id,
+    manifest_interface_instance,
+    retriever,
+    storage,
+    raise_on_error=False,
+    require_empty_layer=True,
+):
+    """ Builds a map containing the digest of each blob referenced by the given manifest,
+        to its associated Blob row in the database. This method also verifies that the blob
+        is accessible under the given repository. Returns None on error (unless raise_on_error
+        is specified). If require_empty_layer is set to True, the method will check if the manifest
+        references the special shared empty layer blob and, if so, add it to the map. Otherwise,
+        the empty layer blob is only returned if it was *explicitly* referenced in the manifest.
+        This is necessary because Docker V2_2/OCI manifests can implicitly reference an empty blob
+        layer for image layers that only change metadata.
+    """
+
+    # Ensure all the blobs in the manifest exist.
+    digests = set(manifest_interface_instance.local_blob_digests)
+    blob_map = {}
+
+    # If the special empty layer is required, simply load it directly. This is much faster
+    # than trying to load it on a per repository basis, and that is unnecessary anyway since
+    # this layer is predefined.
+    if EMPTY_LAYER_BLOB_DIGEST in digests:
+        digests.remove(EMPTY_LAYER_BLOB_DIGEST)
+        blob_map[EMPTY_LAYER_BLOB_DIGEST] = get_shared_blob(EMPTY_LAYER_BLOB_DIGEST)
+        if not blob_map[EMPTY_LAYER_BLOB_DIGEST]:
+            if raise_on_error:
+                raise CreateManifestException("Unable to retrieve specialized empty blob")
+
+            logger.warning("Could not find the special empty blob in storage")
+            return None
+
+    if digests:
+        query = lookup_repo_storages_by_content_checksum(repository_id, digests, with_uploads=True)
+        blob_map.update({s.content_checksum: s for s in query})
+        for digest_str in digests:
+            if digest_str not in blob_map:
+                logger.warning(
+                    "Unknown blob `%s` under manifest `%s` for repository `%s`",
+                    digest_str,
+                    manifest_interface_instance.digest,
+                    repository_id,
+                )
+
+                if raise_on_error:
+                    raise CreateManifestException("Unknown blob `%s`" % digest_str)
+
+                return None
+
+    # Special check: If the empty layer blob is needed for this manifest, add it to the
```

Comment:
Just for my understanding, in Docker v2 the empty layer is not needed?

Reply:
Correct. Its only needed for Docker V2 Schema 1